### DISCUSSION TITLE
Trouble inserting GeometryPart into an iModel

### DISCUSSION THREAD

**Discussion by spowell-216** [iTwin/itwinjs-core#2947] [Q&A] [answered]
Can someone please explain what i am doing wrong?  I am trying to insert a geometryPart into a newly created/downloaded iModel from the iModelHub.  I cannot reproduce this error locally when using a SnapshotDb.  

iModelJs ver: 2.19.26

error:  `Bad Element: insertElement with class=BisCore:GeometryPart: error inserting `

this is the method for creating the GeometryPartProps :
``` 
public static recreateGeometry(sProps: GeometryPartProps, tDefinitionModelId: Id64String, iModelDb: IModelDb ): GeometryPartProps {
    const props: GeometryPartProps = {
      classFullName: GeometryPart.classFullName,
      model: tDefinitionModelId,
      code: GeometryPart.createCode(iModelDb, tDefinitionModelId, sProps.code.value ? sProps.code.value : "empty code"),
      geom: sProps.geom,
      bbox: sProps.bbox,  
    }
    return props; 
}
```

this is the GeometryPartProps returned:
```
{
  "classFullName": "BisCore:GeometryPart",
  "model": "0x20000000001",
  "code": { "spec": "0x1", "scope": "0x20000000001", "value": "StyleSymbol-8631" },
  "geom": [{ "header": { "flags": 0 } }, { "textString": { "font": 1, "text": "E", "height": 0.001524, "origin": [-0.0008052431000982602, -0.00040700674451391266, 0] } }],
  "bbox": { "low": [-0.0008052431000982602, -0.0011690067445139125, 0], "high": [0.0006145772124017397, 0.0018789932554860872, 0] }
}
```

i have verified that the model is a newly created definition model using the following: 
`tDefinitionModelId = DefinitionModel.insert(iModel, parentSubjectId, "Definitions");`

i have verified that there is a   `"font": 1`  for the  "geom"

`number: 1, fontProps: {"id":1,"type":1,"name":"Arial"}`

**Comment by bbastings** [comment]
Did you reserve your code value? If you don't require a unique code use Code.createEmpty()...

https://www.itwinjs.org/learning/backend/concurrencycontrol/#code-reservation

Probably not an issue since it worked in a snapshot, but to eliminate the text as the source of the failure, try creating a geometry part with a line segment?

* Also, you don't need to set bbox, that will computed from the geometry automatically, any values you specify are ignored.

**Reply by spowell-216** [reply]
I am not currently reserving code values.  Currently I only implementing the following after  opening the targetdb:
```
      iModelDb.concurrencyControl.setPolicy(new ConcurrencyControl.OptimisticPolicy());
      iModelDb.concurrencyControl.startBulkMode();
```
sourcedb i have no concurrencyControl on, just preforming a `SnapshotDb.openFile`.

I am taking your advice and trying couple different scenarios of creating geometryParts, with various codes and geom.   
and thanks for the reminder about bbox, i've commented that out.

**Reply by spowell-216** [reply]
@bbastings I have gone back and did some more unit tests and I am now able to reproduce the error against a SnapshotDb. 

```
  test("Test inserting geometryPart", () => {
    // insert a nonFont geometry
    const nProp: GeometryPartProps = { 
      classFullName: "BisCore:GeometryPart",
      model: defModelId,
      code: {spec: "0x1", scope: defModelId, value: "NonFont" },
      geom: [
        { "header": { "flags": 0 } },
        { "arc": { "center": [0.002830322975865115, -0.0015032541506855902, 0], "vectorX": [0.003048, 0, 0], "vectorY": [0, 0.003048, 0], "sweepStartEnd": [21.501434324047928, 156.50143432404792] } }  
      ]
    };
    const nonFontId = targetDb.elements.insertElement(nProp);
    expect(nonFontId).toBeDefined();

    // check to see if there are any fonts
    chai.assert(targetDb.fontMap.getFont(1) == undefined);

    // add font 
    const font1: FontProps = { id: 1, type: 1, name: "Aria1"};
    targetDb.fontMap.addFonts([font1]);
    expect(targetDb.fontMap.getFont(1)).toEqual(font1);

    // add geometryPart with font 1
    const gProp1: GeometryPartProps = {
      classFullName: "BisCore:GeometryPart",
      model: defModelId,
      code: { spec: "0x1", scope: defModelId, value: "StyleSymbol-8631" },
      geom:[
        { header: { flags: 0 }}, 
        {textString: {font: 1, text: "E", height: 0.001524, origin: [-0.0008052431000982602, -0.00040700674451391266, 0]}}
      ]};

    const fontId1 = targetDb.elements.insertElement(gProp1);
    expect(fontId1).toBeDefined();
    expect(fontId1).not.toEqual(nonFontId);
    
    // add more fonts
    const font2: FontProps = { id: 2, type:1, name: "Arial Narrow" };
    const font3: FontProps = { id: 3, type:1, name: "Braggadocio" };
    targetDb.fontMap.addFonts([font2, font3]);
    chai.assert(targetDb.fontMap.getFont("Arial Narrow"));
    chai.assert(targetDb.fontMap.getFont(3));

    // add geometryPart with font 3
    const gProp3: GeometryPartProps = {
      classFullName: "BisCore:GeometryPart",
      model: defModelId,
      code: { spec: "0x1", scope: defModelId, value: "StyleSymbol-8632" },
      geom:[
        { header: { flags: 0 }}, 
        {textString: {font: 3, text: "E", height: 0.001524, origin: [-0.0008052431000982602, -0.00040700674451391266, 0]}}
      ]};
    const fontId3 = targetDb.elements.insertElement(gProp3);
    expect(fontId3).toBeDefined();
    expect(fontId3).not.toEqual(fontId1);
  });
```
the test will fails when trying to add a geometryPart with a reference to a font
![image](https://user-images.githubusercontent.com/51784707/146933140-df0b3aeb-1371-4a28-8bd2-7e8109d9cada.png)

the targetDb was initialized with SnapshotDb.createEmpty() with the following models 
```
    const subId = targetDb.elements.getRootSubject().id;
    const defModelId = DefinitionModel.insert(targetDb,subId, "def");
    const infModelId = InformationRecordModel.insert(targetDb, subId, "Info Records");
    const dwgListModelId = DocumentListModel.insert(targetDb, subId, "PID Model");
    const dwgModelId = Drawing.insert(targetDb, dwgListModelId, "Sample DWG");
```

**Comment by bbastings** [comment, marked as answer]
// add font 
    const font1: FontProps = { id: 1, type: 1, name: "Aria1"};
    targetDb.fontMap.addFonts([font1]);

This doesn't actually add a font, it's just adding an entry into a map (which you really shouldn't be able to change directly as the map is populated from the available fonts that are embedded in the imodel). To really add a font you need to use IModelDb.embedFont, which currently is only supported on Windows for true type fonts.

This is an area that is still in need to further development.

**Reply by spowell-216** [reply]
thanks for clarifying, hopefully this gets some attention in the new year.

**Reply by kabentley** [reply]
we are actively working on this. Should be something (in 3.0) in January, I hope.
